### ISSUE TITLE
Warnings on Lambda due to lack of home folder

### ISSUE THREAD

**Issue by nmoutschen** [awslabs/aws-sdk-rust#307]
### Version

```
│   ├── aws-sdk-sts v0.0.25-alpha
├── aws-sdk-dynamodb v0.0.25-alpha
├── aws-sdk-eventbridge v0.0.25-alpha
```

### Platform

```
Darwin <REDACTED> 19.6.0 Darwin Kernel Version 19.6.0: Tue Oct 12 18:34:05 PDT 2021; root:xnu-6153.141.43~1/RELEASE_X86_64 x86_64
```

### Description

When using any crate that depends on `aws_config` on an AWS Lambda function, it emits warnings that it _"could not determine home directory but home expansion was requested"_ and it _"failed to get selected 'default' profile"_. This is due to the fact that Lambda doesn't have a home directory, but uses environment variables to get AWS credentials.

The SDK works fine on Lambda, but this pollutes log entries with unnecessary warning messages, which could cause issues if someone is using tracing to log messages and triggers an alarm on a high number of warning messages in CloudWatch Logs.

__Steps to reproduce__:

1. Deploy a sample application to AWS:

```bash
git clone https://github.com/aws-samples/serverless-rust-demo.git
cd serverless-rust-demo
make
```

2. Look at the CloudWatch Logs for any of the deployed Lambda functions.
3. See an entry such as this in the logs:

```json
{
    "timestamp": "Nov 23 19:11:02.837",
    "level": "WARN",
    "fields": {
        "message": "could not determine home directory but home expansion was requested"
    },
    "target": "aws_config::profile::parser::source",
    "span": {
        "name": "load_config_file"
    },
    "spans": [
        {
            "name": "get_store"
        },
        {
            "name": "load_config_file"
        }
    ]
}
```

**Comment by Velfi** [comment]
I'll see if I can reproduce this. In the meantime,
That message specifically is coming from this function:
```rust
fn expand_home(path: impl AsRef<Path>, home_dir: &Option<String>) -> PathBuf {
    let path = path.as_ref();
    let mut components = path.components();
    let start = components.next();
    match start {
        None => path.into(), // empty path,
        Some(Component::Normal(s)) if s == "~" => {
            // do homedir replacement
            let path = match home_dir {
                Some(dir) => {
                    tracing::debug!(home = ?dir, path = ?path, "performing home directory substitution");
                    dir.clone()
                }
                None => {
                    tracing::warn!(
                        "could not determine home directory but home expansion was requested"
                    );
                    // if we can't determine the home directory, just leave it as `~`
                    "~".into()
                }
            };
            let mut path: PathBuf = path.into();
            // rewrite the path using system-specific path separators
            for component in components {
                path.push(component);
            }
            path
        }
        // Finally, handle the case where it doesn't begin with some version of `~/`:
        // NOTE: in this case we aren't performing path rewriting. This is correct because
        // this path comes from an environment variable on the target
        // platform, so in that case, the separators should already be correct.
        _other => path.into(),
    }
}
```
Is it possible you have an env var set that's using a tilde (`~`) somewhere?

**Comment by nmoutschen** [comment]
Hi @Velfi !

I only set two variables for this Lambda function: `RUST_LOG` and `TABLE_NAME`. Lambda will also set a few more environment variables as defined [in the documentation](https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime).

If it helps you troubleshoot, this is [where I initialize the SDK](https://github.com/aws-samples/serverless-rust-demo/blob/main/src/utils.rs#L13-L27). All I do is `aws_config::load_from_env()`.

**Comment by Velfi** [comment]
I have a fix for this: https://github.com/awslabs/smithy-rs/pull/893
I just have to test it with your code

**Comment by nmoutschen** [comment]
Thanks a lot @Velfi !

Let me take a look at the code to make sure it aligns with how we detect Lambda environments for other libraries.
